File: go/types/indexed_sequence_diff.go
// Copyright 2016 Attic Labs, Inc. All rights reserved.
// Licensed under the Apache License, version 2.0:
// http://www.apache.org/licenses/LICENSE-2.0

package types

func indexedSequenceDiff(last indexedSequence, lastHeight int, lastOffset uint64, current indexedSequence, currentHeight int, currentOffset uint64, changes chan<- Splice, closeChan <-chan struct{}, maxSpliceMatrixSize uint64) bool {
	if lastHeight > currentHeight {
		lastChild := last.(indexedMetaSequence).getCompositeChildSequence(0, uint64(last.seqLen())).(indexedSequence)
		return indexedSequenceDiff(lastChild, lastHeight-1, lastOffset, current, currentHeight, currentOffset, changes, closeChan, maxSpliceMatrixSize)
	}

	if currentHeight > lastHeight {
		currentChild := current.(indexedMetaSequence).getCompositeChildSequence(0, uint64(current.seqLen())).(indexedSequence)
		return indexedSequenceDiff(last, lastHeight, lastOffset, currentChild, currentHeight-1, currentOffset, changes, closeChan, maxSpliceMatrixSize)
	}

	compareFn := last.getCompareFn(current)
	initialSplices := calcSplices(uint64(last.seqLen()), uint64(current.seqLen()), maxSpliceMatrixSize,
		func(i uint64, j uint64) bool { return compareFn(int(i), int(j)) })

	for _, splice := range initialSplices {
		if !isMetaSequence(last) || splice.SpRemoved == 0 || splice.SpAdded == 0 {
			// We have meta data about the number of leaves below us, so if an entire meta sequence was removed, we don't need to dig down to compute the diff, we can just use math.
			lastAtCum := uint64(0)
			if splice.SpAt > 0 {
				lastAtCum = last.cumulativeNumberOfLeaves(int(splice.SpAt) - 1)
			}
			lastEndRemoveCum := uint64(0)
			if splice.SpAt+splice.SpRemoved > 0 {
				lastEndRemoveCum = last.cumulativeNumberOfLeaves(int(splice.SpAt+splice.SpRemoved) - 1)
			}
			currentFromCum := uint64(0)
			if splice.SpFrom > 0 {
				currentFromCum = current.cumulativeNumberOfLeaves(int(splice.SpFrom) - 1)
			}
			currentEndAddedCum := uint64(0)
			if splice.SpFrom+splice.SpAdded > 0 {
				currentEndAddedCum = current.cumulativeNumberOfLeaves(int(splice.SpFrom+splice.SpAdded) - 1)
			}

			splice.SpRemoved = lastEndRemoveCum - lastAtCum
			splice.SpAdded = currentEndAddedCum - currentFromCum
			splice.SpAt = lastOffset + lastAtCum
			if splice.SpAdded > 0 {
				splice.SpFrom = currentOffset + currentFromCum
			}

			select {
			case changes <- splice:
			case <-closeChan:
				return false
			}

		} else {
			lastChild := last.(indexedMetaSequence).getCompositeChildSequence(splice.SpAt, splice.SpRemoved).(indexedSequence)
			currentChild := current.(indexedMetaSequence).getCompositeChildSequence(splice.SpFrom, splice.SpAdded).(indexedSequence)
			lastChildOffset := lastOffset
			if splice.SpAt > 0 {
				lastChildOffset += last.cumulativeNumberOfLeaves(int(splice.SpAt) - 1)
			}
			currentChildOffset := currentOffset
			if splice.SpFrom > 0 {
				currentChildOffset += current.cumulativeNumberOfLeaves(int(splice.SpFrom) - 1)
			}
			if ok := indexedSequenceDiff(lastChild, lastHeight-1, lastChildOffset, currentChild, currentHeight-1, currentChildOffset, changes, closeChan, maxSpliceMatrixSize); !ok {
				return false
			}
		}
	}

	return true
}

File: js/src/indexed-sequence-diff.js
// @flow

// Copyright 2016 Attic Labs, Inc. All rights reserved.
// Licensed under the Apache License, version 2.0:
// http://www.apache.org/licenses/LICENSE-2.0

import type {Splice} from './edit-distance.js';
import {calcSplices, SPLICE_ADDED, SPLICE_AT, SPLICE_FROM,
  SPLICE_REMOVED} from './edit-distance.js';
import {IndexedMetaSequence} from './meta-sequence.js';
import {invariant} from './assert.js';
import type {IndexedSequence} from './indexed-sequence.js';

export function diff(last: IndexedSequence, lastHeight: number, lastOffset: number,
                     current: IndexedSequence, currentHeight: number, currentOffset: number,
                     maxSpliceMatrixSize: number): Promise<Array<Splice>> {

  if (lastHeight > currentHeight) {
    invariant(lastOffset === 0 && currentOffset === 0);
    invariant(last instanceof IndexedMetaSequence);
    return last.getCompositeChildSequence(0, last.length).then(lastChild =>
        diff(lastChild, lastHeight - 1, lastOffset, current, currentHeight, currentOffset,
             maxSpliceMatrixSize));
  }

  if (currentHeight > lastHeight) {
    invariant(lastOffset === 0 && currentOffset === 0);
    invariant(current instanceof IndexedMetaSequence);
    return current.getCompositeChildSequence(0, current.length).then(currentChild =>
        diff(last, lastHeight, lastOffset, currentChild, currentHeight - 1, currentOffset,
             maxSpliceMatrixSize));
  }

  invariant(last.isMeta === current.isMeta);
  invariant(lastHeight === currentHeight);

  const splices = calcSplices(last.length, current.length, maxSpliceMatrixSize,
    last.getCompareFn(current));

  const splicesP = splices.map(splice => {
    if (!last.isMeta || splice[SPLICE_REMOVED] === 0 || splice[SPLICE_ADDED] === 0) {
      // We have meta data about the number of leaves below us, so if an entire meta sequence was
      // removed, we don't need to dig down to compute the diff, we can just use math.
      let lastAtCum = 0;
      if (splice[SPLICE_AT] > 0) {
        lastAtCum = last.cumulativeNumberOfLeaves(splice[SPLICE_AT] - 1);
      }
      let lastEndRemoveCum = 0;
      if (splice[SPLICE_AT] + splice[SPLICE_REMOVED] > 0) {
        lastEndRemoveCum =
            last.cumulativeNumberOfLeaves(splice[SPLICE_AT] + splice[SPLICE_REMOVED] - 1);
      }
      let currentFromCum = 0;
      if (splice[SPLICE_FROM] > 0) {
        currentFromCum = current.cumulativeNumberOfLeaves(splice[SPLICE_FROM] - 1);
      }
      let currentEndAddedCum = 0;
      if (splice[SPLICE_FROM] + splice[SPLICE_ADDED] > 0) {
        currentEndAddedCum =
            current.cumulativeNumberOfLeaves(splice[SPLICE_FROM] + splice[SPLICE_ADDED] - 1);
      }

      splice[SPLICE_REMOVED] = lastEndRemoveCum - lastAtCum;
      splice[SPLICE_ADDED] = currentEndAddedCum - currentFromCum;
      splice[SPLICE_AT] = lastOffset + lastAtCum;
      if (splice[SPLICE_ADDED] > 0) {
        splice[SPLICE_FROM] = currentOffset + currentFromCum;
      }

      return [splice];
    }

    invariant(last instanceof IndexedMetaSequence && current instanceof IndexedMetaSequence);
    const lastChildP = last.getCompositeChildSequence(splice[SPLICE_AT], splice[SPLICE_REMOVED]);
    const currentChildP = current.getCompositeChildSequence(splice[SPLICE_FROM],
      splice[SPLICE_ADDED]);

    let lastChildOffset = lastOffset;
    if (splice[SPLICE_AT] > 0) {
      lastChildOffset += last.cumulativeNumberOfLeaves(splice[SPLICE_AT] - 1);
    }
    let currentChildOffset = currentOffset;
    if (splice[SPLICE_FROM] > 0) {
      currentChildOffset += current.cumulativeNumberOfLeaves(splice[SPLICE_FROM] - 1);
    }

    return Promise.all([lastChildP, currentChildP]).then(childSequences =>
      diff(childSequences[0], lastHeight - 1, lastChildOffset, childSequences[1], currentHeight - 1,
           currentChildOffset,
           maxSpliceMatrixSize));
  });

  return Promise.all(splicesP).then(spliceArrays => {
    const splices = [];
    for (let i = 0; i < spliceArrays.length; i++) {
      splices.push(...spliceArrays[i]);
    }
    return splices;
  });
}
